src/lib.rs

```diff
@@ -0,0 +1,1035 @@
+        }
+    }
+}
+
+impl Allocation {
+    #[inline(always)]
+    pub(crate) fn new(kind: AllocationKind, index: usize) -> Self {
+        assert!(index < (1 << 28));
+        Self {
+            bits: ((kind as u8 as u32) << 29) | (index as u32),
+        }
+    }
+
+    #[inline(always)]
+    pub fn none() -> Allocation {
+        Allocation::new(AllocationKind::None, 0)
+    }
+
+    #[inline(always)]
+    pub fn reg(preg: PReg) -> Allocation {
+        Allocation::new(AllocationKind::Reg, preg.index())
+    }
+
+    #[inline(always)]
+    pub fn stack(slot: SpillSlot) -> Allocation {
+        Allocation::new(AllocationKind::Stack, slot.0 as usize)
+    }
+
+    #[inline(always)]
+    pub fn kind(self) -> AllocationKind {
+        match (self.bits >> 29) & 7 {
+            5 => AllocationKind::None,
+            6 => AllocationKind::Reg,
+            7 => AllocationKind::Stack,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_none(self) -> bool {
+        self.kind() == AllocationKind::None
+    }
+
+    #[inline(always)]
+    pub fn is_reg(self) -> bool {
+        self.kind() == AllocationKind::Reg
+    }
+
+    #[inline(always)]
+    pub fn is_stack(self) -> bool {
+        self.kind() == AllocationKind::Stack
+    }
+
+    #[inline(always)]
+    pub fn index(self) -> usize {
+        (self.bits & ((1 << 28) - 1)) as usize
+    }
+
+    #[inline(always)]
+    pub fn as_reg(self) -> Option<PReg> {
+        if self.kind() == AllocationKind::Reg {
+            Some(PReg::from_index(self.index()))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn as_stack(self) -> Option<SpillSlot> {
+        if self.kind() == AllocationKind::Stack {
+            Some(SpillSlot(self.index() as u32))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn bits(self) -> u32 {
+        self.bits
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: u32) -> Self {
+        debug_assert!(bits >> 29 >= 5);
+        Self { bits }
+    }
+}
+
+// N.B.: These values must be *disjoint* with the values used to
+// encode `OperandPolicy`, because they share a 3-bit field.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum AllocationKind {
+    None = 5,
+    Reg = 6,
+    Stack = 7,
+}
+
+impl Allocation {
+    #[inline(always)]
+    pub fn class(self) -> RegClass {
+        match self.kind() {
+            AllocationKind::None => panic!("Allocation::None has no class"),
+            AllocationKind::Reg => self.as_reg().unwrap().class(),
+            AllocationKind::Stack => self.as_stack().unwrap().class(),
+        }
+    }
+}
+
+/// A helper that wraps either an `Operand` or an `Allocation` and is
+/// able to tell which it is based on the tag bits.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct OperandOrAllocation {
```

Comment:
I guess it would be useful for debuginfo to track where an SSA var can be found.

Reply:
This was developed in Cranelift when I was attempting the straight regalloc2-API port, and then lifted here since its implementation details depend on internal properties of the `Operand` and `Allocation`. The basic use-case is for instruction representations that are mutation-based, holding an `Operand` before allocation and an `Allocation` afterward. Indeed if one already has the `Vec<Allocation>` allocated, and one is designing from scratch, there's not much point to that; but it might be helpful in adapting other compilers, I suppose.

Reply:
The thing is, the current regalloc API has `inst_operands` returning a `&[Operand]`. This requires operands to be in a `Vec<Operand>` or `[Operand; N]`. This doesn't work if the operands are in a `Vec<OperandOrAllocation>`/`[OperandOrAllocation; N]`.

Reply:
Yes, so the way that the Cranelift regalloc2-native-API port works (currently, on the WIP branch) is: the `regalloc2::Function` implementation collects operands into a separate `Vec` (shared amongst all instructions) in order to provide the backing store for the `inst_operands` slices; and when the allocation result is returned, it goes back to the `MachInst`s to edit the `OperandOrAllocation`s in-place.

The reasoning was: given the requirements that (i) `Inst` is standalone, i.e. knows how to emit itself without auxiliary data, (ii) `Inst` is a simple enum with named fields for reg operands, and (iii) we minimize memory usage, I needed some sort of sum type and a preprocessing step (via a visitor pattern) to gather `Operand`s and postprocessing to set `Allocation`s. The bitpacking to make it memory-efficient is dependent on the details of `Operand` and `Allocation` (specifically the non-overlapping "kind" tags) so it made sense to put it into regalloc2 rather than Cranelift.

One might then ask: this requires collecting operands into a separate `Operand` vec first; isn't that inefficient? Actually BTRA does the same (`RegsAndVecs`) in its preprocessing; both designs start with a "visitor" pattern that lets the `Inst` store operands in a heterogenous way, with named fields per enum-arm; this just pushes the preprocessing to the other side of the API boundary.

It's very possible that there is a better way here -- probably the cleanest would be for the `Inst` to store a fixed-size array of `Operand` per enum arm, come up with some macro or codegen magic to try to make this as ergonomic as native `Reg` enum fields today, and then accept a separate `Allocation` slice as an argument to the `emit()` function -- but the intent of the initial port was to try to minimize disruptive design changes all at once. (The WIP diff is ~4k lines already so one can come to various conclusions how successful that will be, but...).

I'm happy to rip this out for now, though, as the Cranelift port to native regalloc2 is still very much up in the air. (It's my next project, FWIW, so I'll be able to spend lots of time thinking about this soon!)

Reply:
[This](https://gist.github.com/Amanieu/ebed6232420ca443dff2bb533bbb7276) is the design I used for my low-level IR. Each `LirInst` has an offset within the operands vector, which continues until the `operands_offset` in the next `LirInst`. After regalloc, the allocations for the operands are fetched directly from the `Vec<Allocation>` in the regalloc output.

Reply:
Right, that's a good general design IMHO. Cranelift was driven by different design considerations as described above (standalone `Inst`, pattern-matching, both symptoms of "no overall context to hold gathered `Operand`s separately). How much this changes in the future is somewhat TBD -- I need to do some work to map out transition steps as I refactor the isel; we probably don't have the luxury of just dropping another large "rewrite the backends" diff that starts from scratch with a new design.